Boston Housing Assignment¶

In this assignment you'll be using linear regression and ridge linear regression to estimate the cost of house in boston, using a well known dataset.

In [1]:
from sklearn import datasets
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cross_validation import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge

In [2]:
bean = datasets.load_boston()
print bean.DESCR

Boston House Prices dataset

Notes
------
Data Set Characteristics:  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive
    
    :Median Value (attribute 14) is usually the target

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pupil-teacher ratio by town
      

In [3]:
def load_boston():
    scaler = StandardScaler()
    boston = datasets.load_boston()
    X=boston.data
    y=boston.target
    X = scaler.fit_transform(X)
    return train_test_split(X,y)

In [4]:
X_train, X_test, y_train, y_test = load_boston()

In [5]:
X_train.shape

(379L, 13L)

In [6]:
linreg = LinearRegression()
linreg.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [7]:
zip (y_test, linreg.predict(X_test))

[(7.0, -6.3432331249439144),
 (44.799999999999997, 37.583260510153572),
 (14.4, 2.9016434511177742),
 (12.300000000000001, 12.632332625889909),
 (13.800000000000001, 5.0769035360500787),
 (36.399999999999999, 32.55388681613772),
 (22.699999999999999, 25.057386515405611),
 (28.699999999999999, 25.356535733304071),
 (19.300000000000001, 21.572712993471303),
 (29.800000000000001, 32.421017030172024),
 (18.5, 19.801188575789389),
 (17.800000000000001, 16.145436553682607),
 (27.899999999999999, 20.784079773923921),
 (19.600000000000001, 19.602023846002155),
 (13.4, 11.908562004324843),
 (20.0, 23.067363476581704),
 (11.9, 8.5555848517828128),
 (25.0, 25.676800167100819),
 (50.0, 41.75320068552297),
 (11.9, 22.795139089271455),
 (20.600000000000001, 17.281249330731473),
 (21.699999999999999, 22.991318161760329),
 (29.899999999999999, 31.329154011825523),
 (32.0, 33.338637356744471),
 (25.0, 22.960915207580836),
 (17.199999999999999, 13.229086983514907),
 (31.699999999999999, 33.0113120020928

In [8]:
y_LnrPred=linreg.predict(X_test)

In [9]:
msValue=mean_squared_error(y_test,y_LnrPred)

In [10]:
msValue

27.847274593280591

In [11]:
rScore=r2_score(y_test,y_LnrPred)
rScore

0.71167533116540715

### MSE and R^2 after linear prediction is:

MSE: 27.847274593280591
r^2: 0.71167533116540715


In [12]:
ridge=Ridge(alpha=0.01)

In [13]:
ridge.fit(X_train,y_train)

Ridge(alpha=0.01, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [14]:
zip(y_test,ridge.predict(X_test))

[(7.0, -6.3426455080225601),
 (44.799999999999997, 37.582903599099694),
 (14.4, 2.9023102195867239),
 (12.300000000000001, 12.632200977910493),
 (13.800000000000001, 5.0768510974579186),
 (36.399999999999999, 32.55420972295174),
 (22.699999999999999, 25.056925062718001),
 (28.699999999999999, 25.356622607089029),
 (19.300000000000001, 21.572581291188914),
 (29.800000000000001, 32.420858301594727),
 (18.5, 19.801061375272074),
 (17.800000000000001, 16.145827803105739),
 (27.899999999999999, 20.783305546562065),
 (19.600000000000001, 19.602337551896156),
 (13.4, 11.909024833256806),
 (20.0, 23.06723933259131),
 (11.9, 8.5549552904087136),
 (25.0, 25.67629191988334),
 (50.0, 41.753194791910587),
 (11.9, 22.795102786696013),
 (20.600000000000001, 17.282130763655012),
 (21.699999999999999, 22.991189813016714),
 (29.899999999999999, 31.32884111114803),
 (32.0, 33.338392123529445),
 (25.0, 22.960877999338326),
 (17.199999999999999, 13.228604289860225),
 (31.699999999999999, 33.010911563700446

In [15]:
y_RegPred=ridge.predict(X_test)

In [16]:
r2RegScore=r2_score(y_test,y_RegPred)

In [17]:
r2RegScore

0.71168428363259373

In [18]:
mseRegScore=mean_squared_error(y_test,y_RegPred)

In [19]:
mseRegScore

27.846409936735455

### MSE and R^2 after ridge linear prediction is:

MSE: 27.846409936735455
r^2: 0.71168428363259373